File: src/services/auth.service.ts
/**
 * Servicio de autenticación.
 * Gestiona login, refresh y validación de perfil instalador.
 */
import apiClient from './api-client';
import { saveToken, removeToken } from '@utils/secure-storage';

export interface LoginRequest {
  username: string;
  password: string;
}

export interface LoginResponse {
  token: string;
  expires_at: string;
  user: {
    id: number;
    nombre: string;
    perfil: string; // 'coach' | 'colaborador'
    merchan: string;
  };
}

// La API envuelve todo en { success: true, data: {...} }
interface ApiResponse<T> {
  success: boolean;
  data: T;
  timestamp: string;
}

export const authService = {
  /**
   * Autenticar usuario contra la API Evora.
   * Valida que el perfil sea coach o colaborador.
   */
  async login(credentials: LoginRequest): Promise<LoginResponse> {
    try {
      const response = await apiClient.post<ApiResponse<LoginResponse>>('/auth/login', credentials);
      const data = response.data.data; // Desempaquetar { success, data: {...} }

      if (data.user.perfil !== 'coach' && data.user.perfil !== 'colaborador') {
        throw new Error('PERFIL_NO_AUTORIZADO');
      }

      await saveToken(data.token);
      return data;
    } catch (err: unknown) {
      // Distinguir errores HTTP del backend
      if (err && typeof err === 'object' && 'response' in err) {
        const axiosErr = err as { response?: { status: number; data?: { error?: { message?: string } } } };
        if (axiosErr.response?.status === 403) {
          const msg = axiosErr.response.data?.error?.message || '';
          if (msg.includes('deshabilitada')) throw new Error('CUENTA_BLOQUEADA');
          throw new Error('PERFIL_NO_AUTORIZADO');
        }
        if (axiosErr.response?.status === 429) {
          throw new Error('DEMASIADOS_INTENTOS');
        }
      }
      throw err;
    }
  },

  /**
   * Renovar token sin re-login.
   */
  async refreshToken(): Promise<string> {
    const response = await apiClient.post<ApiResponse<{ token: string; expires_at: string }>>('/auth/refresh');
    const data = response.data.data;
    await saveToken(data.token);
    return data.token;
  },

  /**
   * Cerrar sesión y eliminar token del dispositivo.
   */
  async logout(): Promise<void> {
    await removeToken();
  },
};

File: lib/services/auth_service.dart
import 'package:dio/dio.dart';
import '../core/api_client.dart';
import '../core/secure_storage.dart';
import '../models/user.dart';

class LoginResponse {
  final String token;
  final String expiresAt;
  final User user;

  const LoginResponse({
    required this.token,
    required this.expiresAt,
    required this.user,
  });
}

/// Servicio de autenticación.
class AuthService {
  /// Autenticar usuario contra la API Evora.
  /// Valida que el perfil sea coach o colaborador.
  static Future<LoginResponse> login({
    required String username,
    required String password,
  }) async {
    try {
      final response = await apiClient.post('/auth/login', data: {
        'username': username,
        'password': password,
      });

      final data = response.data['data'] as Map<String, dynamic>;
      final user = User.fromJson(data['user'] as Map<String, dynamic>);

      if (user.perfil != 'coach' && user.perfil != 'colaborador') {
        throw AuthException('PERFIL_NO_AUTORIZADO');
      }

      await SecureStorage.saveToken(data['token'] as String);
      return LoginResponse(
        token: data['token'] as String,
        expiresAt: data['expires_at'] as String,
        user: user,
      );
    } on DioException catch (e) {
      if (e.response?.statusCode == 403) {
        final msg = e.response?.data?['error']?['message'] as String? ?? '';
        if (msg.contains('deshabilitada')) {
          throw AuthException('CUENTA_BLOQUEADA');
        }
        throw AuthException('PERFIL_NO_AUTORIZADO');
      }
      if (e.response?.statusCode == 429) {
        throw AuthException('DEMASIADOS_INTENTOS');
      }
      rethrow;
    }
  }

  /// Renovar token sin re-login.
  static Future<String> refreshToken() async {
    final response = await apiClient.post('/auth/refresh');
    final data = response.data['data'] as Map<String, dynamic>;
    final token = data['token'] as String;
    await SecureStorage.saveToken(token);
    return token;
  }

  /// Cerrar sesión.
  static Future<void> logout() async {
    await SecureStorage.removeToken();
  }
}

class AuthException implements Exception {
  final String code;
  const AuthException(this.code);

  @override
  String toString() => 'AuthException: $code';
}
